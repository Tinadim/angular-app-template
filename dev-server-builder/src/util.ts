import { BuilderOutput } from '@angular-devkit/architect';
import { logging } from '@angular-devkit/core';
import * as path from 'path';
import * as rxjs from 'rxjs';
import * as browserSync from 'browser-sync';
import * as url from 'url';
import { delay, ignoreElements, startWith, switchMap, tap } from 'rxjs/operators';
import { getAvailablePort, spawnAsObservable } from '@nguniversal/builders/src/ssr-dev-server/utils';
import { createProxyMiddleware } from 'http-proxy-middleware';
import { Options } from './interfaces';

const IGNORED_STDOUT_MESSAGES = [
    'server listening on',
    'Angular is running in the development mode. Call enableProdMode() to enable the production mode.'
];

export function getClientWorkspaceDetails(options: Options) {
    const { browserWorkspaceRoot } = options;
    if (path.isAbsolute(browserWorkspaceRoot)) {
        return browserWorkspaceRoot;
    }
    return path.resolve(process.cwd(), browserWorkspaceRoot);
}

export function startNodeServer(serverOutput: BuilderOutput, port: number, logger: logging.LoggerApi) {
    const outputPath = serverOutput.outputPath as string;
    const serverPath = path.join(outputPath, 'main.js');
    const env = Object.assign(Object.assign({}, process.env), { PORT: '' + port });
    return rxjs.of(null)
        .pipe(
            delay(0), // Avoid EADDRINUSE error since it will cause the kill event to be finish.
            switchMap(() => spawnAsObservable('node', [`"${serverPath}"`], { env, shell: true })),
            tap(({ stderr, stdout }) => {
                if (stderr) {
                    logger.error(stderr);
                }
                if (stdout && !IGNORED_STDOUT_MESSAGES.some((x: string) => stdout.includes(x))) {
                    logger.info(stdout);
                }
            }),
            ignoreElements(),
            // Emit a signal after the process has been started
            startWith(undefined)
        );
}

export async function initBrowserSync(
    browserSyncInstance: browserSync.BrowserSyncInstance,
    nodeServerPort: number,
    options: Options,
) {
    if (browserSyncInstance.active) {
        return browserSyncInstance;
    }
    const { port: browserSyncPort, open, host, publicHost } = options;
    const bsPort = browserSyncPort || await getAvailablePort();
    const bsOptions: { [key: string]: any } = {
        proxy: {
            target: `localhost:${nodeServerPort}`,
            proxyOptions: {
                xfwd: true
            },
            proxyRes: [
                (proxyRes: any) => {
                    if ('headers' in proxyRes) {
                        proxyRes.headers['cache-control'] = undefined;
                    }
                },
            ]
            // proxyOptions is not in the typings
        },
        host,
        port: bsPort,
        ui: false,
        server: false,
        notify: false,
        ghostMode: false,
        logLevel: 'silent',
        open,
    };
    const publicHostNormalized = publicHost && publicHost.endsWith('/')
        ? publicHost.substring(0, publicHost.length - 1)
        : publicHost;

    if (publicHostNormalized) {
        const { protocol, hostname, port, pathname } = url.parse(publicHostNormalized);
        const defaultSocketIoPath = '/browser-sync/socket.io';
        const defaultNamespace = '/browser-sync';
        const hasPathname = Boolean(pathname && pathname !== '/');
        const namespace = hasPathname ? pathname + defaultNamespace : defaultNamespace;
        const path = hasPathname ? pathname + defaultSocketIoPath : defaultSocketIoPath;
        bsOptions.socket = {
            namespace,
            path,
            domain: url.format({
                protocol,
                hostname,
                port,
            }),
        };
        // When having a pathname we also need to create a reverse proxy because socket.io
        // will be listening on: 'http://localhost:4200/ssr/browser-sync/socket.io'
        // However users will typically have a reverse proxy that will redirect all matching requests
        // ex: http://testinghost.com/ssr -> http://localhost:4200 which will result in a 404.
        if (hasPathname) {
            // Remove leading slash
            bsOptions.scriptPath = (p: string) => p.substring(1);
            bsOptions.middleware = [
                createProxyMiddleware(defaultSocketIoPath, {
                    target: url.format({
                        protocol: 'http',
                        hostname: host,
                        port: bsPort,
                        pathname: path,
                    }),
                    ws: true,
                    logLevel: 'silent',
                }),
            ];
        }
    }
    return new Promise((resolve, reject) => {
        browserSyncInstance.init(bsOptions, (error, bs) => {
            if (error) {
                reject(error);
            }
            else {
                resolve(bs);
            }
        });
    });
}

export function getBaseUrl(bs: browserSync.BrowserSyncInstance) {
    return `${bs.getOption('scheme')}://${bs.getOption('host')}:${bs.getOption('port')}`;
}

export function mapErrorToMessage(error: any) {
    if (error instanceof Error) {
        return error.message;
    }
    if (typeof error === 'string') {
        return error;
    }
    return '';
}
