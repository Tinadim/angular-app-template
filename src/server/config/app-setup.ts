import express, { Application, Request, Response } from 'express';
import bodyParser from 'body-parser';
import Router from 'express-promise-router';
import { ngExpressEngine } from '@nguniversal/express-engine';
import { APP_BASE_HREF } from '@angular/common';
import { resolve } from 'path';
import { mount } from '../api';
import { AppServerModule } from '../server-side-rendering/app-server-module';

const distFolder = resolve(__dirname, '../../../dist/client');
const useSsr = process.env.USE_SSR;
const renderer = useSsr
    ? (req: Request, res: Response) => res.render('index', { req, providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }] })
    : (req: Request, res: Response) => res.sendFile(`${distFolder}/index.html`);

export function buildApp() {
    const app = express();
    setupMiddleware(app);
    bootEndpoints(app);
    serveStaticFiles(app);
    return app;
}

function serveStaticFiles(app: Application) {
    app.get('*.*', express.static(distFolder, {
        maxAge: '1y'
    }));

    app.get('*', renderer);
}

function setupMiddleware(app: Application) {
    if (useSsr) {
        app.engine('html', ngExpressEngine({
            bootstrap: AppServerModule,
        }));
    }

    app.set('view engine', 'html');
    app.set('views', distFolder);
    app.use(bodyParser.urlencoded({ extended: true }));
    app.use(bodyParser.json());
}

function bootEndpoints(app: Application) {
    const router = Router();
    mount(router);
    app.use('/api', router);
}
