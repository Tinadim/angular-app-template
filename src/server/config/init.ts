import 'zone.js/dist/zone-node';
import { resolve } from 'path';
import { config as configEnv } from 'dotenv-safe';

const example = resolve(process.cwd(), '.env.example');
configEnv({ example });

console.log('<<<<< USE_SSR', process.env.USE_SSR, '>>>>>');
