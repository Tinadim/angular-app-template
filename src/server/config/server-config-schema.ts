import { $enum } from 'ts-enum-util';

enum EnvVars {
    NodeEnv = 'NODE_ENV',
    Port = 'PORT',
    UseShutDownHandler = 'USE_SHUTDOWN_HANDLER',
    UseSsr = 'USE_SSR'
}

enum ArgVars {
    Port = 'port',
    UseShutDownHandler = 'useShutdownHandler',
    UseSsr = 'useSsr'
}

export const ENV_VARS = $enum(EnvVars).getValues();

export const ARG_VARS = $enum(ArgVars).getValues();

export const DEFAULTS = {
    port: 8080,
    useSsr: true,
    useShutdownHandler: false
};

export const envSchema: object = {
    properties: {
        [EnvVars.Port]: {
            type: 'number',
            default: DEFAULTS.port,
        },
        [EnvVars.UseShutDownHandler]: {
            type: 'boolean',
            default: DEFAULTS.useShutdownHandler,
        },
        [EnvVars.UseSsr]: {
            type: 'boolean',
            default: DEFAULTS.useSsr,
        }
    },
    type: 'object'
};

export const argvSchema: object = {
    properties: {
        [ArgVars.Port]: {
            type: 'number',
        },
        [ArgVars.UseShutDownHandler]: {
            type: 'boolean',
        },
        [ArgVars.UseSsr]: {
            type: 'boolean',
        }
    },
    type: 'object'
};
