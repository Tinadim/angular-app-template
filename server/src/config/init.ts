import 'zone.js/dist/zone-node';
import { resolve } from 'path';
import * as appRoot from 'app-root-path';
import * as dotenv from 'dotenv';
import { registerProvider } from '@app/client-provider';
import * as provider from '@client/src/client-provider';

dotenv.config({ path: resolve(appRoot.path, '.env') });

console.log('<<<<< ROOT', appRoot.path, '>>>>>');

const resolvedPath = require.resolve('../../../client/src/client-provider');
// const resolvedPath = resolve(__dirname, '../../client/src/client-provider');

console.log('ENV\n\n', JSON.stringify(process.env.USE_SSR));

registerProvider(resolvedPath, provider);
