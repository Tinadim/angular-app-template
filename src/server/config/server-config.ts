import Ajv from 'ajv';
import { camelCase, cloneDeep, mapKeys, pick } from 'lodash';
import { resolve } from 'path';
import { parse as parseArgv } from 'yargs';
import { envSchema, argvSchema, ENV_VARS, ARG_VARS } from './server-config-schema';

export class ServerConfig {
    readonly distFolderPath: string;
    readonly port: number;
    readonly isProduction: boolean;
    readonly useShutdownHandler: boolean;
    readonly useSsr: boolean;

    constructor() {
        this.distFolderPath = resolve(process.cwd(), './dist/client');

        const mergedConfigs = this.mergeConfigs();

        this.isProduction = process.env.NODE_ENV === 'production';
        this.port = mergedConfigs.port;
        this.useSsr = mergedConfigs.useSsr;
        this.useShutdownHandler = mergedConfigs.useShutdownHandler;
    }

    private mergeConfigs() {
        const ajv = new Ajv({
            allErrors: true,
            coerceTypes: true,
            format: 'full',
            useDefaults: true,
        });
        const relevantEnvVars = pick(process.env, ENV_VARS);
        const argvCopy = pick(cloneDeep(parseArgv()), ARG_VARS);
        this.validateEnv(ajv, relevantEnvVars);
        this.validateArgv(ajv, argvCopy);

        const mappedEnvVars = mapKeys(relevantEnvVars, (value: any, key: string) => camelCase(key));
        return Object.assign({}, mappedEnvVars, argvCopy) as any;
    }

    private validateEnv(ajv: Ajv.Ajv, envVars: object) {
        const valid = ajv.validate(envSchema, envVars);
        if (!valid) {
            throw this.formatAjvError(ajv.errors);
        }
    }

    private validateArgv(ajv: Ajv.Ajv, argVars: object) {
        const valid = ajv.validate(argvSchema, argVars);
        if (!valid) {
            throw this.formatAjvError(ajv.errors);
        }
    }

    private formatAjvError(errors: Ajv.ErrorObject[]) {
        // TODO add better error formatting
        throw new Ajv.ValidationError(errors);
    }
}

export const serverConfig = new ServerConfig();
