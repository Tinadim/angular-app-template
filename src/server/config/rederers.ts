import { Request, RequestHandler, Response } from 'express';
import { APP_BASE_HREF } from '@angular/common';
import { serverConfig } from './server-config';

const { useSsr, distFolderPath } = serverConfig;

export function renderIndex(req: Request, res: Response) {
    res.render('index', {
        req,
        providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
}

export function sendIndex(req: Request, res: Response) {
    res.sendFile(`${distFolderPath}/index.html`);
}

export function getRenderer(): RequestHandler {
    return useSsr ? renderIndex : sendIndex;
}
