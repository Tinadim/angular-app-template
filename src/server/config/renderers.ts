import { Request, RequestHandler, Response } from 'express';
import { APP_BASE_HREF } from '@angular/common';
import { serverConfig } from './server-config';

export function renderIndex(req: Request, res: Response) {
    res.render('index', {
        req,
        providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }],
    });
}

export function sendIndex(req: Request, res: Response) {
    const { distFolderPath } = serverConfig;
    res.sendFile(`${distFolderPath}/index.html`);
}

export function getRenderer(): RequestHandler {
    return serverConfig.useSsr ? renderIndex : sendIndex;
}
