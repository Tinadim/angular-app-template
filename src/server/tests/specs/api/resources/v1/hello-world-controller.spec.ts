import request from 'supertest';
import { Application } from 'express';
import { testError } from '@test-lib/util';
import { buildApp } from '@server/config/app-setup';

describe('hello-world-controller.spec.ts', () => {
    const app: Application = buildApp();

    describe('[POST] /api/v1/hello-world', () => {
        function sendRequest(data: object = {}) {
            const url = '/api/v1/hello-world';
            return request(app)
                .post(url)
                .send(data);
        }

        it('Should return a 400 error if the required parameters are not provided', async () => {
            const response = await sendRequest({});
            expect(response.status).toBe(400);
            testError(response.body.error);
        });

        it('Should return a 400 error if the parameter is not a string', async () => {
            const response = await sendRequest({ sender: '' });
            expect(response.status).toBe(400);
            testError(response.body.error);
        });

        it('Should return a 200 response', async () => {
            const response = await sendRequest({ sender: 'foo' });
            expect(response.status).toBe(200);
            expect(response.body).toMatchSnapshot();
        });
    });
});
