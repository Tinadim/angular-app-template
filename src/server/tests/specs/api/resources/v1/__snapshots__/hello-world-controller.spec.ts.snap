// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello-world-controller.spec.ts [POST] /api/v1/hello-world Should return a 200 response 1`] = `
Object {
  "data": Object {
    "sender": "foo",
  },
  "success": true,
}
`;

exports[`hello-world-controller.spec.ts [POST] /api/v1/hello-world Should return a 400 error if the parameter is not a string 1`] = `
Object {
  "code": "invalidParameters",
  "errors": Array [
    Object {
      "code": "invalidParameters",
      "field": "body.sender",
      "message": "should NOT be shorter than 1 characters",
    },
  ],
  "message": "Invalid parameters provided",
}
`;

exports[`hello-world-controller.spec.ts [POST] /api/v1/hello-world Should return a 400 error if the required parameters are not provided 1`] = `
Object {
  "code": "invalidParameters",
  "errors": Array [
    Object {
      "code": "missingRequiredParameter",
      "field": "body",
      "message": "should have required property 'sender'",
    },
  ],
  "message": "Invalid parameters provided",
}
`;
